sympy/utilities/tests/test_wester.py

```diff
@@ -2181,21 +2182,24 @@ def test_U4():
     assert diff(x**n, x, n).rewrite(factorial) == factorial(n)
 
 
-@XFAIL
-def test_U5():
-    # https://github.com/sympy/sympy/issues/6681
-    # f(g(x)).diff(x,2) returns Derivative(g(x), x)**2*Subs(Derivative(
-    #  f(_xi_1), _xi_1, _xi_1), (_xi_1,), (g(x),)) + Derivative(g(x), x, x)*
-    #  Subs(Derivative(f(_xi_1), _xi_1), (_xi_1,), (g(x),))
-    raise NotImplementedError("f(g(t)).diff(t,2) Subs not performed")
+def test_U5():  # XXX check
+    # issue 6681
+    t = symbols('t')
+    a1 = (
+        Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
+        Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2)
+    a2 = (
+        Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
+        Derivative(g(t), t)*Derivative(f(g(t)), g(t), t))
```

Comment:
This doesn't look right. I don't think there should be a derivative wrt g(t).

Reply:
Here is my hand derivation:

```
f(g(t)).diff(t)

Derivative(f(g(t)),g(t))*Derivative(g(t),t)

identify as a*b, then (a*b).diff(t) is

Derivative(a, t)*b + a*Derivative(b, t)

Substituting a and b in derivatives

Derivative(Derivative(f(g(t)),g(t)), t)*b + a*Derivative(Derivative(g(t),t), t)

and simplifying

Derivative(f(g(t)), g(t), t)*b + a*Derivative(g(t), t, 2)

then substituting in a and b for the remaining occurrences gives a2

Derivative(f(g(t)), g(t), t)*Derivative(g(t),t) + Derivative(f(g(t)), g(t))*Derivative(g(t), t, 2)
\--------------------------------------------/

The first terms differ in a2 and a1 (below)

/---------------------------------------------------\
Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2 + Derivative(f(g(t)), g(t))*Derivative(g(t), t, 2)

Checking if the indicated expressions are plausibly equal

>>> f=lambda x:x**4
... g=lambda x:exp(x)
... 

from a2

>>> Subs(Derivative(f(x), x), x, g(t)).diff(t)*Derivative(g(t),t).doit()
12*exp(4*t)

from a1

>>> (Subs(Derivative(f(x),x,2),x,g(t))*(g(t).diff(t))**2).doit()
12*exp(4*t)

Now check all of a2

>>> (Subs(Derivative(f(x), x), x, g(t)).diff(t)*Derivative(g(t),t) + Subs(Derivative(f(x), x),x,g(t))*Derivative(g(t), t, 2)).doit()
16*exp(4*t)

and compare to original

>>> f(g(t)).diff(t,2)
16*exp(4*t)
```

Reply:
a = Derivative(f(g(t)), g(t)) is short for Derivative(f(y), y).subs(y, g(t)), or Df(g(t)) if Df(y) is used to denote Derivative(f(y), y).

Then Derivative(a, t) is Derivative(Df(g(t)), g(t)))*Derivative(g(t), t) where
Derivative(Df(g(t)), g(t)) means Derivative(Df(y), y).subs(y, g(t)) which is
Derivative(f(y), (y, 2)).subs(y, g(t)).

I am not sure what Derivative(f(g(t)), g(t), t)) should mean but it looks different.

Reply:
Perhaps it's better to prefer that the evaluated form be represented with Subs:

```
>>> Derivative(f(g(x)),g(x))
Derivative(f(g(x)), g(x))
>>> _.doit()
Subs(Derivative(f(_xi_0), _xi_0), _xi_0, g(x))

Here's the expression in question:

>>> f(g(x)).diff(x,2)
Derivative(g(x), x)**2*Subs(Derivative(f(_xi_1), (_xi_1, 2)), _xi_1,
g(x)) + Derivative(g(x), (x, 2))*Subs(Derivative(f(_xi_1), _xi_1),
_xi_1, g(x))

A particular case of f and g:

>>> f(g(x)).diff(x,2).subs(f,Lambda(x,x**4)).subs(g,Lambda(x,exp(x))).doit()
16*exp(4*x)
```

Reply:
> I am not sure what Derivative(f(g(t)), g(t), t)) should mean but it looks different.

Are you mainly objecting to this representation? It means what you have said: `Subs(f(y).diff(y), y, g(t)).diff(t)`. I *think* that `Derivative(f(g(t)), g(t), t)).doit()` will spell this out explicitly but will check on this later.

Reply:
> Subs(f(g(t)).diff(t), g(t), y) is 0 since you are differentiating wrt t.

This is not necessarily 0, if `g(t)` depends on `t`. For example, if `f(y) = y**2`, then `f(g(t)).diff(t)` is `2*g(t)*g(t).diff(t)` which becomes `2*y*g(t).diff(t)`.
This is the reason why I would like to ban derivative expressions having both `g(t)` and `t`.

Reply:
Here's why I think it's always zero:

```python
>>> f(g(t)).diff(t)
Derivative(g(t), t)*Subs(Derivative(f(_xi_1), _xi_1), (_xi_1,), (g(t),))
```
When you replace g(t) with y that first factor will be 0 and that forces the whole expression to 0. This is true for your example, too, since `Subs(g(t).diff(t), g(t), y)` will be 0.

I'm very keen to get this right, however, and we are already doing something that Ronan had warned against but others felt that the "function as symbol" semantics for derivation was important sugar to keep. So please continue to point out notation that you think is ambiguous. I am especially interested to know if there is actually a case where the order of differentiation matters. I think for our current assumptions that it does not.

Reply:
> When you replace g(t) with y that first factor will be 0

I think that g(t) cannot be replaced *inside a derivative*. That is bound to give wrong results.

Reply:
I'm not sure what meaning `Subs(g(t).diff(t), g(t), y)` can have: `g(t).diff(t)` no longer depends on `g(t)`. This is the current assumption of derivatives wrt functions: `f(x).diff(x,f(x)) = 0`. Derivatives of the same expression to different orders are independent of each other hence f(x) (and order 0 derivative) and f(x).diff(x) (a first derivative) are independent. This is the assumption on which derivatives in SymPy are defined.

Reply:
This is the note that appears in Derivative's docstring:

```
    When differentiation wrt a non-Symbol is attempted, the non-Symbol
    is temporarily converted to a Symbol while the differentiation
    is performed::

        >>> eq = x + f(x)**2
        >>> eq.diff(f(x)).subs(f(x), y)
        2*y
        >>> eq.subs(f(x), y).diff(y)
        2*y

    This means, also, that all functions are treated as independent
    of other functions and are not related to their free symbols::

        >>> x.diff(f(x))
        0
        >>> g(x).diff(f(x))
        0

    It also means that derivatives of functions are assumed to be
    independent of the functions themselves and derivatives of
    different orders::

        >>> from sympy import Derivative
        >>> F = f(x)
        >>> dF = Derivative(F, x)
        >>> d2F = Derivative(F, x, 2)
        >>> F.diff(dF)
        0
        >>> dF.diff(F)
        0
        >>> dF.diff(d2F)
        0
```

Reply:
It seems that SymPy will evaluate `a1` and `a2` in the same way.
```
>>> from sympy.abc import t
>>> f, g = symbols('f g', cls=Function)
>>> a1 = (
...         Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
...         Derivative(f(g(t)), (g(t), 2))*Derivative(g(t), t)**2)
>>> a2 = (
...         Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) +
...         Derivative(g(t), t)*Derivative(f(g(t)), g(t), t))
>>> a1.doit() == a2.doit()
True
```
So I take back my doubts and agree that both `a1` and `a2` are correct. (I should have done this test in the first place, sorry.)

However, I also get
```
>>> f(g(t)).diff(t,2) in (a1, a2)
False
>>> f(g(t)).diff(t,2).doit() == a1.doit()
True
```
Perhaps we could dispense with `a2`?

Reply:
Those were the results from master. In this branch I get `a2`.
```
>>> f(g(t)).diff(t, 2)
Derivative(f(g(t)), g(t))*Derivative(g(t), (t, 2)) + Derivative(g(t), t)*Derivative(f(g(t)), g(t), t)
```
Would it be hard to make this `a1`? (Automatically processing higher derivatives wrt both `g(t)` and `t`?)

Reply:
The reason it doesn't get processed is because all processing past the first non-symbol is halted. I asked on the mailing list for examples where the ordering of g(t) and t matters. Aaron has said that with our assumptions it doesn't matter. So I will modify the sorting routine to move all symbols to the front of the queue. That should take care of this.